expression/builtin_string.go

```diff
@@ -2945,23 +2986,20 @@ func (b *builtinFormatWithLocaleSig) Clone() builtinFunc {
 // evalString evals FORMAT(X,D,locale).
 // See https://dev.mysql.com/doc/refman/5.7/en/string-functions.html#function_format
 func (b *builtinFormatWithLocaleSig) evalString(row chunk.Row) (string, bool, error) {
-	x, isNull, err := b.args[0].EvalString(b.ctx, row)
-	if isNull || err != nil {
-		return "", true, err
-	}
-
-	d, isNull, err := b.args[1].EvalString(b.ctx, row)
+	x, d, isNull, err := evalNumDecArgsForFormat(b, row)
```

Comment:
Actually, the target to use different `sig`s for different input types is to reduce the branch predicate in the execution phase, to gain the same improvement as the JIT method, i.e:
```go
if arg0.GetType().EvalType() == types.ETDecimal {
	...
} else {
	...
}
```

In the execution phase, the type is a constant, so the above branch predicate can be eliminated to avoid CPU branch predication and exhibit the power of CPU pipeline.

Reply:
For the first branch in line 2922, we cannot eliminate it since we need the exact type to construct `builtinFunc`? for the second branch in line 2950, we cannot eliminate it either?

Reply:
Yes, the first branch in line 2922 can not be eliminated. But the second branch in line 2950 can be eliminated. Because in this time, `arg0.GetType().EvalType()` is definitive if we create separate signature for `FORMAT()` for different input argument types, like:
```
builtinFormatDecimalWithLocaleSig -- FORMAT(decimal, int, string)
builtinFormatRealWithLocaleSig    -- FORMAT(float64, int, string)
```

Reply:
Oh I got it, but the shortcoming of different `sig`s is leading to a lot of redundant/duplicate code? The number of `format` arguments can be 2 or 3, while the first argument can be decimal or double type, so we have 4 combinations of `sig`s, it may look ugly IMHO... Since this function would not be called frequently, I prefer readability over eliminating branch predication for performance here?

Reply:
> it may look ugly

Can't agree more! In the near future, expression evaluation will be refactored to the vectorized version. And as you said "this function would not be called frequently", It's fine to keep the current implementation for readability.